src/osd/PG.cc

```diff
@@ -1335,7 +1337,32 @@ void PG::calc_replicated_acting(
       !primary->second.is_incomplete() &&
       primary->second.last_update >=
         auth_log_shard->second.log_tail) {
-    ss << "up_primary: " << up_primary << ") selected as primary" << std::endl;
+    if (HAVE_FEATURE(osdmap->get_up_osd_features(), SERVER_NAUTILUS)) {
+      auto approx_objects_missing =
+        primary->second.stats.stats.sum.num_objects_missing;
+      auto auth_version = auth_log_shard->second.last_update.version;
+      auto primary_version = primary->second.last_update.version;
+      if (auth_version > primary_version) {
+        approx_objects_missing += auth_version - primary_version;
+      } else {
+        approx_objects_missing += primary_version - auth_version;
+      }
+      auto force_threshold = cct->_conf.get_val<uint64_t>(
+        "osd_force_auth_primary_missing_objects");
```

Comment:
Making choose_acting depend on a config variable is dangerous because it can lead to a cycle if the value differs between OSDs (with acting toggling back and forth between two primaries with different settings).  We do it with the ignore_les option but that has led to confusion/problems in the field as a result.

Reply:
Yeah. That's why I introduce a very strict constraint above:
```
if (HAVE_FEATURE(osdmap->get_up_osd_features(), SERVER_NAUTILUS)) 
```
And choosing acting under the control of some configurable option is something that the async_recovery code already does.. see:

**osd_async_recovery_min_pg_log_entries**